src/java.base/share/classes/jdk/internal/misc/Blocker.java

```diff
@@ -0,0 +1,143 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.misc;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.concurrent.ForkJoinPool;
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+/**
+ * Defines static methods to mark the beginning and end of a possibly blocking
+ * operation. The methods are intended to be used with try-finally as follows:
+ * {@snippet lang=java :
+ *     long comp = Blocker.begin();
+ *     try {
+ *         // blocking operation
+ *     } finally {
+ *         Blocker.end(comp);
+ *     }
+ * }
+ * If invoked from a virtual thread and the underlying carrier thread is a
+ * CarrierThread then the code in the block runs as if it were in run in
+ * ForkJoinPool.ManagedBlocker. This means the pool can be expanded to support
+ * additional parallelism during the blocking operation.
+ */
+public class Blocker {
+    private static final JavaLangAccess JLA;
+    static {
+        JLA = SharedSecrets.getJavaLangAccess();
+        if (JLA == null) {
+            throw new InternalError("JavaLangAccess not setup");
+        }
+    }
+
+    private Blocker() { }
+
+    private static Thread currentCarrierThread() {
+        return JLA.currentCarrierThread();
+    }
+
+    /**
+     * Marks the beginning of possibly blocking operation.
+     * @return the return value from the attempt to compensate
+     */
+    public static long begin() {
+        if (VM.isBooted()
+                && currentCarrierThread() instanceof CarrierThread ct && !ct.inBlocking()) {
+            ct.beginBlocking();
+            long comp = ForkJoinPools.beginCompensatedBlock(ct.getPool());
+            assert currentCarrierThread() == ct;
+            return comp;
+        }
+        return 0;
+    }
+
+    /**
+     * Marks the beginning of possibly blocking operation.
+     * @param blocking true if the operation may block, otherwise false
+     * @return the return value from the attempt to compensate when blocking is true,
+     * another value when blocking is false
+     */
+    public static long begin(boolean blocking) {
+        return (blocking) ? begin() : 0;
+    }
+
+    /**
+     * Marks the end an operation that may have blocked.
```

Comment:
I think this should have been "Marks the end of an ..."

Reply:
Yes, there's a typo there - thanks.